qiskit/circuit/library/generalized_gates/unitary.py

```diff
@@ -185,14 +190,19 @@ def control(
         if not annotated:
             mat = self.to_matrix()
             cmat = _compute_control_matrix(mat, num_ctrl_qubits, ctrl_state=None)
-            iso = Isometry(cmat, 0, 0)
+            from qiskit.synthesis.unitary.qsd import qs_decomposition
+
+            cmat_def = qs_decomposition(cmat, opt_a1=True, opt_a2=False)
+            if not matrix_equal(Operator(cmat_def).to_matrix(), cmat, atol=1e-7):
+                self.definition = Isometry(cmat, 0, 0).definition
```

Comment:
code updated to use Isometry the calculated Operator is different than the original matrix.

Reply:
I don't think there are special numerical errors. QSD is done recursively and this is the reason that we get numeric errors. However, Isometry is not iterative. I can add similar tests to Isometry.